partiql-logical-planner/src/typer.rs

```diff
@@ -0,0 +1,893 @@
+                } else if derived_type.is_unordered_collection() {
+                    PartiqlType::new_bag(BagType::new(Box::new(ty)))
+                } else {
+                    self.errors.push(TypingError::IllegalState(format!(
+                        "Expecting Collection for the output Schema but found {:?}",
+                        &ty
+                    )));
+                    ty
+                };
+
+                // Marking the output Schema in Typing Environment:
+                //
+                // The clauses of an PartiQL SFW query are evaluated in the following order:
+                // WITH, FROM, LET, WHERE, GROUP BY, HAVING, LETTING (which is special to PartiQL),
+                // ORDER BY, LIMIT / OFFSET
+                // and SELECT (or SELECT VALUE or PIVOT, which are both special to ion PartiQL).
+                // -- PartiQL Spec. 2019 Section 3.3:
+                self.type_env_stack.push(ty_ctx![(
+                    &ty_env![(string_to_sym(OUTPUT_SCHEMA_KEY), schema)],
+                    derived_type
+                )]);
+            }
+            BindingsOp::Sink => {
+                if let Some(ty) =
+                    self.retrieve_type_from_local_ctx(&string_to_sym(OUTPUT_SCHEMA_KEY))
+                {
+                    self.output = Some(ty);
+                }
+            }
+            _ => self.errors.push(TypingError::NotYetImplemented(format!(
+                "Unsupported BindingOperator: {:?}",
+                &op
+            ))),
+        }
+    }
+
+    fn type_vexpr(&mut self, v: &ValueExpr, lookup_order: LookupOrder) {
+        fn to_bindings(s: &StructType) -> Vec<(SymbolPrimitive, PartiqlType)> {
+            s.fields()
+                .into_iter()
+                .map(|field| {
+                    let sym = SymbolPrimitive {
+                        value: field.name().to_string(),
+                        case: CaseSensitivity::CaseInsensitive,
+                    };
+
+                    (sym, field.ty().clone())
+                })
+                .collect()
+        }
+
+        fn binding_to_sym(binding: &BindingsName) -> SymbolPrimitive {
+            match binding {
+                BindingsName::CaseSensitive(s) => SymbolPrimitive {
+                    value: s.to_string(),
+                    case: CaseSensitivity::CaseSensitive,
+                },
+                BindingsName::CaseInsensitive(s) => SymbolPrimitive {
+                    value: s.to_string(),
+                    case: CaseSensitivity::CaseInsensitive,
+                },
+            }
+        }
+
+        match v {
+            ValueExpr::VarRef(binding_name, _) => {
+                let sym = binding_to_sym(binding_name);
+
+                match lookup_order {
+                    LookupOrder::GlobalLocal => {
+                        let ty = self.resolve_global_then_local(&sym);
+                        let ty = self.element_type(&ty);
```

Comment:
I don't think the unnesting should be part of the `Scan` mainly b/c of the possibility of having local lookups in the `Scan`. However perhaps it makes more sense to push this down to `resolve_global` since this is the only place that requires to unnest.